### PR TITLE
Remove pending notifications

### PR DESCRIPTION
**Pull request by scottinet**
:warning: depends on https://github.com/kuzzleio/kuzzle/pull/1445 and https://github.com/kuzzleio/kuzzle-plugin-cluster/pull/84

# Description

Remove the `state` property from subscription options and from real-time notifications.
That property was deprecated since Kuzzle 1.5.0.

edit: SDKs don't use that `state` property, nor do they document it in their `subscribe` API, so this does not break applications using them

### CODE REVIEW

File: lib/api/core/models/repositories/tokenRepository.js [hunk 1]
```diff
@@ -221,6 +221,7 @@ class TokenRepository extends Repository {
       emptyKeyLength = super.getCacheKey('').length,
       userKey = super.getCacheKey(userId + '#*');
 
+    // @todo ask @scottinet or @bvidis for this
```
Comment (kuzzle) [inline lib/api/core/models/repositories/tokenRepository.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/controllers/adminController.js [hunk 2]
```diff
@@ -84,15 +80,20 @@ class AdminController extends BaseController {
...
 
-    const
-      database = request.input.args.database,
-      cacheEngine = this.kuzzle.services.list[database];
+    const database = request.input.args.database;
+
+    let cacheEngine;
 
-    if (!cacheEngine || typeof cacheEngine.flushdb !== 'function') {
+    // @todo allow only publicCache
```
Comment (kuzzle) [inline lib/api/controllers/adminController.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/models/repositories/repository.js [hunk 3]
```diff
@@ -59,19 +60,28 @@ class Repository {
...
-    if (options.databaseEngine === null) {
-      this.databaseEngine = options.databaseEngine;
+    // if set to null, it means that we explicitely dont want a indexEngine for this repository
+    // as null is falsy, without this check, the indexEngine is always set...
+    if (options.indexEngine === null) {
+      this.indexEngine = options.indexEngine;
     }
     else {
-      this.databaseEngine = options.databaseEngine || this.kuzzle.internalEngine;
+      // @todo refactor storageEngine set (it's dirty to pass it with options.indexEngine)
```
Comment (kuzzle) [inline lib/api/core/models/repositories/repository.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/storage/indexCache.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    if (collection && ! this.indexes[index].collections.includes(collection)) {
+      this.indexes[index].collections.push(collection);
+
+      modified = true;
+    }
+
+    if (modified) {
+      // @todo modify cluster for scope
```
Comment (kuzzle) [inline lib/api/core/storage/indexCache.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/core/plugins/pluginContext.js [hunk 5]
```diff
@@ -75,23 +76,67 @@ class PluginContext {
...
+        get: () => {
+          return {
+            bootstrap: () => pluginIndexEngine.init(),
+            createCollection: (...args) => pluginIndexEngine.createCollection(...args)
+          };
+        }
+      });
+
+      /**
+       * @todo Use a CollectionEngine class instead repository
```
Comment (kuzzle) [inline lib/api/core/plugins/pluginContext.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

File: lib/api/controllers/indexController.js [hunk 6]
```diff
@@ -161,7 +161,8 @@ class IndexController extends BaseController {
    * @returns {Promise<Object>}
    */
   refreshInternal() {
-    return this.kuzzle.internalEngine.refresh()
+    // @todo remove refresh* methods
```
Comment (kuzzle) [inline lib/api/controllers/indexController.js]: INFO Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

**Comment by kuzzle** [comment]
SonarQube analysis reported 8 issues
* ![INFO][INFO] 8 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [collectionController.js#L70](https://github.com/kuzzleio/kuzzle/blob/d1fd0a7b05117033a907ae6303f4a93454490678/lib/api/controllers/collectionController.js#L70): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)
1. ![INFO][INFO] [kuzzle.js#L53](https://github.com/kuzzleio/kuzzle/blob/d1fd0a7b05117033a907ae6303f4a93454490678/lib/api/kuzzle.js#L53): Complete the task associated to this TODO comment. [rule](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1135)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: lib/api/core/notifier.js [hunk 7]
```diff
@@ -133,22 +133,15 @@ class NotifierController {
    *
    * @param {Request} request
    * @param {string} scope - Scope of the notification
```
Comment (Aschen) [inline lib/api/core/notifier.js]: Forgotten comment

File: lib/api/core/notifier.js [hunk 8]
```diff
@@ -161,26 +154,16 @@ class NotifierController {
...
-        rooms,
-        request,
-        'in',
-        'done',
-        'create',
-        {
-          _source: newDocument._source,
-          _id: newDocument._id
-        })
+      .notifyDocument(rooms, request, 'in', 'create', {
```
Comment (Aschen) [inline lib/api/core/notifier.js]: I also prefer this way to wrap method with an object as last argument :)

Comment (scottinet) [reply lib/api/core/notifier.js]: I'm still playing with some styles, but this one seems neat. I'm considering adding that rule to our coding style documentation.

File: lib/api/core/hotelClerk.js [hunk 9]
```diff
@@ -344,20 +343,17 @@ class HotelClerk {
...
-    if (request.input.args.state && ['all', 'done', 'pending'].indexOf(request.input.args.state) === -1) {
-      errorsManager.throw('internal', 'hotelclerk', 'incorrect_state_value', request.input.args.state);
-    }
-    channel.state = request.input.args.state || 'done';
+    const
+      channel = {},
+      allowed = ['all', 'in', 'out', 'none'];
 
-    if (request.input.args.scope && ['all', 'in', 'out', 'none'].indexOf(request.input.args.scope) === -1) {
+    if (request.input.args.scope && allowed.indexOf(request.input.args.scope) === -1) {
```
[the remark this answers is not in the document]
Comment (scottinet) [reply lib/api/core/hotelClerk.js]: I didn't apply this suggestion as-is: `null` and `undefined` are perfectly acceptable values. ;-)

**Pull request merged**